### PR TITLE
Exception handling items

### PR DESCRIPTION
**Pull request by PeterAntal**
-Perform null check and add exception logging in RuleSet.
-Explicitly recognize connection termination on ServerOrder.Deserialize()
-Use explicit exception type for SDL2HardwareCursor failures.

### CODE REVIEW

File: OpenRA.Game/GameRules/ActorInfo.cs [hunk 1]
```diff
@@ -38,17 +38,17 @@ public ActorInfo(ObjectCreator creator, string name, MiniYaml node)
 			{
 				Name = name;
 
-				var abstractActorType = name.StartsWith("^");
+				var traitCollectionNamePrefix = "^";
+				var ignoreMissingFields = name.StartsWith(traitCollectionNamePrefix);
```
Comment (pchote) [inline OpenRA.Game/GameRules/ActorInfo.cs]: Trying to parse an `ActorInfo` from a partial template is bogus to begin with.  A tricker but IMO much cleaner fix would be to filter these out before trying to construct them in the `Ruleset`.

**Comment by pchote** [comment]
Don't worry about the Appveyor failures.  That will be fixed by #14855.

File: OpenRA.Game/GameRules/ActorInfo.cs (hunk 1 above)
Comment (PeterAntal) [reply OpenRA.Game/GameRules/ActorInfo.cs]: Hmm, yes - this was a conservative change in that regard- I'm definitely still building up my mental model here.

If I'm interpreting correctly (a very big if, given my ignorance of the range of data shapes travelling here) you would like to recognize and pre-emptively filter the incoming list of Actors names in RuleSet.Load(), for generating the info's?

I'll do some quick prelim checks to get a sense on if that approach hits any obvious snags/problem cases.

Comment (PeterAntal) [reply OpenRA.Game/GameRules/ActorInfo.cs]: I've prepped a change based on suggestion above - if that approach looks sound, I'll drop the failure handling from here.
https://github.com/OpenRA/OpenRA/pull/14859

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs [hunk 2]
```diff
@@ -244,9 +250,12 @@ public SDL2HardwareCursor(Size size, byte[] data, int2 hotspot)
 
 					// This call very occasionally fails on Windows, but often works when retried.
 					for (var retries = 0; retries < 3 && Cursor == IntPtr.Zero; retries++)
+					{
```
Comment (abcdefg30) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: No braces around single-line statements, please.

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs [hunk 3]
```diff
@@ -189,7 +189,7 @@ public IHardwareCursor CreateHardwareCursor(string name, Size size, byte[] data,
 			}
 			catch (Exception ex)
 			{
-				throw new InvalidDataException("Failed to create hardware cursor `{0}` - {1}".F(name, ex.Message), ex);
+				throw new SDL2HardwareCursorException("Failed to create hardware cursor `{0}` - {1}".F(name, ex.Message), ex);
```
Comment (pchote) [inline OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: What does a custom exception type win us here?

File: OpenRA.Game/Server/ServerOrder.cs [hunk 4]
```diff
@@ -42,6 +42,10 @@ public static ServerOrder Deserialize(BinaryReader r)
 						return new ServerOrder(name, data);
 					}
 
+				case 0xbf:
+					Console.WriteLine("Connection Terminated - This isn't a server order.");
```
Comment (pchote) [inline OpenRA.Game/Server/ServerOrder.cs]: Can we make this message consistent with the `0xff` case?

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs (hunk 3 above)
Comment (PeterAntal) [reply OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: Naming Accuracy: Failures I've observed here I've seen were not caused by invalid inputs to SDL.

Most instances of InvalidDataException I've seen in the codebase are pertinent to issues of incorrect data formats, where this one does not appear to be actionable by consumer in that sense, with SDL not rejecting the *format* of data it is supplied in callstacks I have been seeing, and the outer loop consuming it successfully retries on repeated attempt, in the current impl(has smell of a race condition).

I prefer to be able to keep my debugger catching first chance exceptions for anomalous paths, while omitting noisy Exception types which are not indicative of local implementation bugs. Where possible, I don't like overlooking exceptions which have a significant probability of indicating actionable malfunctions.

The cursor failure is a recurrent, but inconsistent repro on my windows environment, for which an InvalidDataException does not seem(to me) to be an accurate reflection of the problem from SDL. Based on threads from other consumers, I'm tempted to follow up on root cause fix with latest debug build of underlying SDL implementation as a separate issue, but that's the context.

File: OpenRA.Game/Server/ServerOrder.cs (hunk 4 above)
Comment (PeterAntal) [reply OpenRA.Game/Server/ServerOrder.cs]: In that case, I'll move it to execute with 0xff.

File: OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs (hunk 3 above)
Comment (PeterAntal) [reply OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: I'm ok with dropping the specific exception types here, if that's a concern on this PR.

Comment (pchote) [reply OpenRA.Platforms.Default/Sdl2GraphicsDevice.cs]: No, that sounds reasonable.  I haven't had time to take a closer look at this PR yet.

File: OpenRA.Game/GameRules/Ruleset.cs [hunk 5]
```diff
@@ -227,10 +227,13 @@ static bool AnyFlaggedTraits(ModData modData, List<MiniYamlNode> actors)
...
 						var traitName = traitNode.Key.Split('@')[0];
 						var traitType = modData.ObjectCreator.FindType(traitName + "Info");
-						if (traitType.GetInterface("ILobbyCustomRulesIgnore") == null)
+						if (traitType != null && traitType.GetInterface("ILobbyCustomRulesIgnore") == null)
 							return true;
 					}
-					catch { }
+					catch (Exception ex)
+					{
+						Log.Write("debug", ex.ToString());
```
Comment (RoosterDragon) [inline OpenRA.Game/GameRules/Ruleset.cs]: Consider `"Error in AnyFlaggedTraits\r\n" + ex.ToString()` so there's some context.

**Review by RoosterDragon — DISMISSED** [review]

File: OpenRA.Game/GameRules/Ruleset.cs (hunk 5 above)
Comment (PeterAntal) [reply OpenRA.Game/GameRules/Ruleset.cs]: Fair point for the logging.

 (Although it does potentially suggest a pattern of an identifier param paired with an exception param, given how many codepaths simply log exception strings)

**Comment by abcdefg30** [comment]
Is the title/description of the first commit still accurate after you split some changes out into other PRs?
While we're here: Can you squash the fixup commit, please?

**Comment by PeterAntal** [comment]
Squashed commits and updated commit description to reflect current payload.

**Review by pchote — APPROVED** [review]

**Pull request merged**